Fifth app - make an image slider

violentr opened the pull request:
Make an api call to get list of products
Create an image slider
  - Has behaviour on click change product pic
  - Working on project tree, split components by its responsibility

Render props components what has no `this.state = {basket: []}` (example)
move it to `/components`
Components who has own `this.state = {}` move stay in `/containers`
Functions what was shared across app should be in `/shared`

- Made major class names refactoring, better code semantics
https://hackernoon.com/structuring-projects-and-naming-components-in-react-1261b6e18d76
- Separate class responsibilities/concern
https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0

violentr committed:
Move Layout Catalog to components folder

violentr committed:
Add npm superagent, make requests

violentr committed:
Add get list of products by calling api, render it

componentDidMount lifecycle

violentr committed:
Add Render list on Products page

use context to make saved state
available to individual product

violentr committed:
Add state component ProductContainer, has products

Received product list from API saved to the
state, context used to populate products
including individual product page

violentr committed:
Add Used productContext to populate list of products

violentr committed:
Better code syntax for Product page

violentr committed:
Add shared helper for the image attrs

violentr committed:
Remove not used code

violentr committed:
Add component Gallery, show extra images for product

violentr committed:
Add Individual product page has gallery of images

violentr committed:
Add Css styles, remove inline styling for Gallery

violentr committed:
Add Show big picture when clicked on product

violentr committed:
Add Big image and 2 additional next to it

Use Portal for lending element in correct
place

index01d commented on Week-1/src/components/Catalog.js:
```diff
@@ -0,0 +1,22 @@
+import React, { Fragment } from 'react'
+import Product from '~/src/ProductCard/Product.js'
+import AddToBasket from '~/src/ProductCard/AddToBasket.js'
+import { image_attributes } from '~/src/shared/helper.js'
```
предпочтительно использовать camelCase вместо snake_case в именах по всему js-проекту

index01d commented on Week-1/src/components/views/Product/index.js:
```diff
@@ -1,17 +1,41 @@
+  <div>
+    <ProductConsumer>
+      { (productContext) => {
+        let products = productContext.products
```
здесь презентационная логика отображения информации смешалась с контейнерной логикой по выборке продукта, можно выделить
views/Product/index.js - контейнерный компонент с выборком
views/Product/Product.js - презентационный компонент с отображением UI

index01d commented on Week-1/src/containers/ProductContainer.js:
```diff
@@ -0,0 +1,44 @@
+import React, { Fragment, Component } from 'react'
+import Catalog from '~/src/components/Catalog.js'
+import request from 'superagent'
+const APIURL = "https://my-json-server.typicode.com/violentr/fake_api/products"
```
нужно чтобы везде были одинаковые стандарты по кавычкам (либо одинарные, либо двойные), чтобы за этим следить можно установить в проект ESLint

index01d replied:
пример конфигурации https://gist.github.com/index01d/76a90487406c518405af94b9d92cf2e8

лучше написать конфигурацию самостоятельно: чуть затратно по времени, зато потом удобно

index01d commented on Week-1/src/containers/ProductContainer.js:
```diff
@@ -0,0 +1,44 @@
+    })
+  }
+  
+  find_by_id(products, id){
```
для единообразия лучше сделать camelCase

index01d reviewed (changes requested):
Предлагаю также поработать над деревом проекта:
1) Установим, что все компоненты должны находится в одной общей директории (components),  ProductCard нужно перенести из src в эту директорию
2) Внутри директории components можно выстроить дерево так
- shared // реиспользуемые, разделяемые между страницами компоненты
- views // компоненты частных страниц и их зависимости
-- Products
--- index.js // компоненты частной страницы
--- e.g.: Catalog.js // компонент-зависимость, который используется только на странице списка продуктов
- Layout.js // особняком компонент общей разметки страницы

Задача состоит в том, чтобы четко разделить и сгруппировать компоненты по ответственности

violentr committed:
Rename var from snake to camel case

violentr committed:
Add Use single quotes accross the project

violentr committed:
Replace snake_case method with camel_case

violentr committed:
Add ProductWrapper, split logic between context and view

violentr committed:
Add Make DOM element mounted well

if this is missing then by changing
to different route gave an error
Error:
Failed to execute 'removeChild' on 'Node':
The node to be removed is not a child of this node.

violentr committed:
Add render props components moved to /components

Pure components - components who has no state and
has

violentr committed:
Add Fix code after moving components

violentr committed:
Add prop-types, validation tool

violentr committed:
Add Portal shared component

violentr committed:
Add Fix code relevant to Portal component

violentr committed:
Add Catalog validate with prop-types

violentr committed:
Add addToBasket validated with prop-types

violentr committed:
Add Price validated with prop-types

violentr committed:
Add Image validated with prop-types

violentr committed:
Add Product validated with prop-types

violentr committed:
Add TextBox validated with prop-types

violentr committed:
Add ProductWrapper validated with prop-types

violentr committed:
Add ProductPage validated with prop-types

violentr committed:
Add CartContainer validated with prop-types

violentr committed:
Add ProductContainer validated with prop-types

violentr committed:
Add Portal validated with prop-types

violentr committed:
Add Render image only when product has image

Avoid empty image blocks

violentr committed:
Add corect code indentation

violentr committed:
Add cosmetic changes, better semantics

violentr committed:
Rename attribute from snake to camel case

violentr committed:
Add move Catalog to views/Products

violentr committed:
Rename ProductCard to Product

Following this
https://hackernoon.com/structuring-projects-and-naming-components-in-react-1261b6e18d76

violentr committed:
Add better component naming concept

Major class names refactoring
following advice at
https://hackernoon.com/structuring-projects-and-naming-components-in-react-1261b6e18d76

violentr committed:
Move shared components to components/shared

violentr committed:
Move Gallery to components/shared

violentr committed:
Add custom settings for eslint

violentr committed:
Eslint do not flag on React components in JSX
